ospfd/ospf_asbr.c

```diff
@@ -305,3 +307,895 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+			 * If a summary route need to be originated but where
+			 * summary route already exist in lsdb with maxage, then
+			 * it need to be refreshed.
+			 */
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: It is already with Maxage, So refresh it (%s/%d)",
+					__func__, inet_ntoa(aggr->p.prefix),
+					aggr->p.prefixlen);
+
+			ospf_external_lsa_refresh(ospf, lsa, ei_aggr,
+						  LSA_REFRESH_FORCE, 1);
+
+			SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+
+		} else {
+
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: Originate Aggregate LSA (%s/%d)",
+					__func__, inet_ntoa(aggr->p.prefix),
+					aggr->p.prefixlen);
+
+			/* Originate summary LSA */
+			lsa = ospf_external_lsa_originate(ospf, ei_aggr);
+			if (lsa)
+				SET_FLAG(aggr->flags,
+					 OSPF_EXTERNAL_AGGRT_ORIGINATED);
+		}
+	}
+}
+
+static void ospf_handle_external_aggr_update(struct ospf *ospf)
+{
+	struct route_node *rn = NULL;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Process modified aggregators.\n", __func__);
+
+	for (rn = route_top(ospf->rt_aggr_tbl); rn; rn = route_next(rn)) {
+		struct ospf_external_aggr_rt *aggr;
+		struct ospf_lsa *lsa = NULL;
+		struct as_external_lsa *asel = NULL;
+		struct external_info ei_aggr;
+		route_tag_t tag = 0;
+
+		if (!rn->info)
+			continue;
+
+		aggr = rn->info;
+
+		if (aggr->action == OSPF_ROUTE_AGGR_DEL) {
+			aggr->action = OSPF_ROUTE_AGGR_NONE;
+			ospf_external_aggr_delete(ospf, rn);
+
+			if (OSPF_EXTERNAL_RT_COUNT(aggr))
+				hash_clean(
+					aggr->match_extnl_hash,
+					(void *)ospf_aggr_handle_external_info);
+
+			hash_free(aggr->match_extnl_hash);
+			XFREE(MTYPE_OSPF_EXTERNAL_RT_AGGR, aggr);
+
+		} else if (aggr->action == OSPF_ROUTE_AGGR_MODIFY) {
+
+			aggr->action = OSPF_ROUTE_AGGR_NONE;
+
+			/* Prepare the extrenal_info for aggregator */
+			memset(&ei_aggr, 0, sizeof(struct external_info));
+			ei_aggr.p = aggr->p;
+			ei_aggr.tag = aggr->tag;
+			ei_aggr.type = 0;
+			ei_aggr.instance = ospf->instance;
+			ei_aggr.route_map_set.metric = -1;
+			ei_aggr.route_map_set.metric_type = -1;
+
+			/* Check if tag modified */
+			if (CHECK_FLAG(aggr->flags,
+				       OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+				lsa = ospf_external_info_find_lsa(ospf,
+								  &ei_aggr.p);
+				if (!lsa) {
+					flog_warn(EC_OSPF_LSA_MISSING,
+						  "%s: Could not refresh/originate %s/%d",
+						  __func__,
+						  inet_ntoa(aggr->p.prefix),
+						  aggr->p.prefixlen);
+					continue;
+				}
+
+				asel = (struct as_external_lsa *)lsa->data;
+				tag = (unsigned long)ntohl(
+					asel->e[0].route_tag);
+
+				/* If tag modified , then re-originate the
+				 * route with modified tag details.
+				 */
+				if (tag != ei_aggr.tag) {
+					if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+						zlog_debug(
+							"%s: Route tag changed(old:%d new:%d,So refresh the summary route.(%s/%d)",
+							__func__, tag,
+							ei_aggr.tag,
+							inet_ntoa(
```

Comment:
use `%pI4` to print ipv4 addresses rather than `inet_ntoa`

Reply:
done